common/quotas/dynamicratelimiter.go

```diff
@@ -24,48 +24,103 @@ package quotas
 
 import (
 	"context"
+	"math"
+	"sync/atomic"
+	"time"
 
 	"golang.org/x/time/rate"
 
 	"github.com/uber/cadence/common/clock"
 )
 
+const _ttl = time.Second * 5
+const _minBurst = 1
+
 // DynamicRateLimiter implements a dynamic config wrapper around the rate limiter,
 // checks for updates to the dynamic config and updates the rate limiter accordingly
 type DynamicRateLimiter struct {
-	rps RPSFunc
-	rl  *RateLimiter
+	rps            RPSFunc
+	rl             clock.Ratelimiter
+	timeSource     clock.TimeSource
+	ttl            time.Duration
+	lastUpdateTime atomic.Value
```

Comment:
I think we have a generic version of this in go.uber.org/atomic or similar

Reply:
Switched to `atomic.Pointer`. `atomic.Value` seems to use a pointer under the hood (since there isn't a way to atomically rewrite that much memory at once) and it gives us generics. There's an `atomic.Time` but it doesn't provide CAS support.

We could alternative use an atomic uint64 for millis since epoch to eliminate the pointers but that's probably negligible from a performance standpoint.

Reply:
yea, perf difference tends to be quite minor, and pointers prevent a LOT of accidentally-racy stuff from being racy (copies, ==, others).  and (strangely) go.uber.org/atomic doesn't embed a `nocopy`, only `nocmp`, so those aren't prevented strongly enough imo.  the safety's a bit more worth it unless that changes.